### ISSUE TITLE
[BUG] Output of ``GroupBy.apply`` is missing the grouping index

### ISSUE THREAD

**Issue by rjzamora** [rapidsai/cudf#14977]
**Describe the bug**
There seems to be a subtle difference between cudf and pandas for `groupby(...).apply(...)`

**Steps/Code to reproduce bug**
```python
import cudf

df = cudf.DataFrame({"a": ["cat", "dog", "cat"], "b": [0, 1, 0]})
df.groupby(["a", "b"]).apply(lambda x: x)
```
```
     a  b
0  cat  0
2  cat  0
1  dog  1
```
**Pandas Result**:
```
           a  b
a   b          
cat 0 0  cat  0
      2  cat  0
dog 1 1  dog  1
```

**Expected behavior**
I would expect the behavior of cudf and pandas to be consistent.

**Additional context**
I found this issue while debugging TPC-h query 16 with dask-cudf + dask-expr.

**Comment by imusa007** [comment]
Looks like this bug is intentional to avoid redundancy. groupby.apply documentation has a similar example of this behavior. 

https://docs.rapids.ai/api/cudf/stable/user_guide/api_docs/api/cudf.core.groupby.groupby.groupby.apply/#cudf.core.groupby.groupby.GroupBy.apply 

So I am not sure if it's a bug or a design choice for better performance or any other reason??

**Comment by rjzamora** [comment]
Thanks for linking that @imusa007! I did see a few notes in the code about this after raising the issue.

It seems okay to close this issue and mark it as "wontfix" for now. Especially if the "fix" requires significant time from anyone (I haven't really looked into it yet). With that said, this divergence will require us to implement/maintain a separate code path in dask-cudf for `nunique`. So, we will need to add code either way.

**Comment by shwina** [comment]
> So I am not sure if it's a bug or a design choice for better performance or any other reason??

@imusa007 I believe the reason was that we couldn't figure out the rules pandas uses to decide whether or not to return the index. If we can pin down exactly when to return the index versus not, we can/should match pandas behaviour.

FYI: @mroeschke just in case you happen to know where to look in the pandas code base/documentation.

**Comment by mroeschke** [comment]
Yeah unfortunately pandas doesn't have a clear rule for `groupby.apply` what type of index is returned when. Some factor that determine the resulting index is the use of `as_index` & whether the UDF returns scalars/same shape as input/mutates input.

A user tried to bravely document the various scenarios but I am not sure how up to date this is: https://github.com/pandas-dev/pandas/issues/22545#issue-355718906
